Mysql replication extended tests

mnagy opened the pull request:
Adding extended tests for MySQL replication extended tests. Heavy WIP. Will need to move some functionality from the test to a separate helper file. @mfojtik PTAL.

@rhcarvalho happy :christmas_tree:

mnagy commented:
This includes a temporary crash fix. @markturansky should fix that soon.

mfojtik commented:
[test][extended:default]

rhcarvalho commented on test/extended/fixtures/nfs-pv-provider.json:
```diff
@@ -0,0 +1,235 @@
+            "storage": "${NFS_PV_CAPACITY}"
+        },
+        "accessModes": [ "ReadWriteOnce" ],
+        "nfs": {
```
@mnagy you can try to use a hostPath instead of NFS, it might remove a bit of complexity.
This is what I documented for PostgreSQL https://github.com/openshift/postgresql/tree/master/examples/replica#deployment after talking with @pmorie

markturansky commented:
working on it.

markturansky commented:
@mnagy here's the PR that will fix what you fixed in secret.go https://github.com/openshift/origin/pull/4384

mfojtik commented on hack/test-extended/default/run.sh:
```diff
@@ -16,9 +16,8 @@ source ${OS_ROOT}/hack/common.sh
-set +e
-
 compile_extended
+set +e
```
this will be fixed soon :-)

mfojtik commented on test/extended/fixtures/nfs-pv-provider.json:
```diff
@@ -0,0 +1,235 @@
+  },
+  "parameters": [
+    {
+      "name": "NAMESPACE",
```
PROJECT_NAME

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,311 @@
+
+// getPods looks up pods with "oc get" applying a label selector and returns
+// the names of matching pods
+func getPods(oc *exutil.CLI, selector string) ([]string, error) {
```
rework this to use `client.PodInterface` instead of exutil.CLI. Then do a client call to get the pods you want.

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,311 @@
+
+// pollForPods will call getPods in a loop until the required number of pods
+// is returned
+func pollForPods(oc *exutil.CLI, selector string, count int, timeout time.Duration) ([]string, error) {
```
this should have following signature:

```go
type PollFunction func(interval, timeout time.Duration, condition ConditionFunc) error
WaitForPods(c client.PodInterface, poll PollFunction) {}
```

jhadvig replied:
should this function be placed in  `framework.go` or a standalone helper so it can be reused in other tests ?

mfojtik replied:
it shoud, but first it has to be more generic to be useful in other test cases...

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,311 @@
+	return pods, err
+}
+
+type mysql struct {
```
how different will the postgresql struct be? If not so much, make it more generic. Same for the functions below.

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,311 @@
+}
+
+// IsUp pings the MySQL server
+func (m *mysql) IsUp() (bool, error) {
```
This should be `WaitForMySQL(oc *exutil.CLI, podName string)`

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,311 @@
+func (m *mysql) RunMysql(root bool, sql string) (string, error) {
+	if root {
+		return m.oc.Run("exec").ExecBashArgs(m.TargetPod, "mysql -h "+"127.0.0.1"+" -uroot -e \""+sql+"\" "+m.Database).Verbose().Output()
+	} else {
```
nuke this else

mfojtik replied:
split this to two functions:

QueryAsRoot()
QueryAsUser()

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,311 @@
+}
+
+// TODO: Generalize, move to exutil
+func (m *mysql) WaitForMysqlOutput(timeout time.Duration, root bool, sql, resultSubstr string) error {
```
we should wait for mysql to accept queries and then expect the query succeed. Not repeat the same query and wait for it to succeed. If the mysql server is up and responding to queries, the queries should succeed (or fail).

mnagy replied:
This is to test replication. If you write to master it might not be replicated in an instant.

mfojtik replied:
Then add another wait function that will exercise just that (WaitForSlaves ?) Don't hide this intent into this function. That is very bad practice as someone not aware of this intent might rewrite the function in future.

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,311 @@
+	Describe("MySQL image replication", func() {
+		JustBeforeEach(func() {
+			// FIXME: Shouldn't do this globaly
+			err := oc.AsAdmin().Run("patch").Args("scc", "privileged", "-p", "{\"users\":[\"system:serviceaccount:openshift-infra:build-controller\",\"system:serviceaccount:default:router\",\"system:serviceaccount:"+oc.Namespace()+":default\"]}").Verbose().Execute()
```
Move this to `SetupNFS` helper

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,311 @@
+		})
+
+		AfterEach(func() {
+			err := oc.AsAdmin().Run("delete").Args("pods", "--all").Verbose().Execute()
```
move this to CleanupNFS helper

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,311 @@
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		It("should work", func() {
```
beautiful name :-)

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,311 @@
+			err = oc.Run("create").Args("-f", configFile).Execute()
+			Expect(err).NotTo(HaveOccurred())
+
+			By("seeing the mysql-master endpoint is up")
```
By("waiting for mysql-master endpoint")

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,311 @@
+			Expect(err).NotTo(HaveOccurred())
+
+			// Get pod names
+			podNames, err := getPods(oc, "name=mysql-master")
```
this is very ugly.

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,311 @@
+
+			// Get pod names
+			podNames, err := getPods(oc, "name=mysql-master")
+			Expect(err).NotTo(HaveOccurred())
```
you will have to prefix **all** Ginkgo/Gomega functions with `g.` and `o.`. See the builds test.

mfojtik replied:
also avoid 'dot' imports (already fixed in other tests).

mfojtik commented on test/extended/util/cli.go:
```diff
@@ -35,10 +37,19 @@ type CLI struct {
 	kubeFramework   *e2e.Framework
 }
 
+type ExitError struct {
```
godoc

mfojtik commented on test/extended/util/cli.go:
```diff
@@ -35,10 +37,19 @@ type CLI struct {
+}
+
+func (e ExitError) Error() string {
+	return fmt.Sprintf("Program exited with status %d", e.ExitStatus)
```
why is this needed? You can just add exit code the the Verbose() output. (or to the default output when the error occurs)

mfojtik replied:
bump?

mnagy replied:
I felt it would be better for the user of the ExitError to be able to extract the value, in case he'd want to ignore a particular error.

mfojtik replied:
no needed anymore, we will panic and convert the panic to `error`

mfojtik commented on test/extended/util/cli.go:
```diff
@@ -226,6 +244,15 @@ func (c *CLI) Args(args ...string) *CLI {
 	return c
 }
 
+func (c *CLI) ExecBashArgs(pod, cmd string) *CLI {
```
godoc

mfojtik replied:
Should have better name... ExecInShell() or something

mfojtik commented on test/extended/util/cli.go:
```diff
@@ -242,7 +269,19 @@ func (c *CLI) Output() (string, error) {
-	return trimmedOutput(c.stdout), err
+	out := trimmedOutput(c.stdout)
+	if c.execBashArgs {
+		re := regexp.MustCompile("@@@([0-9]+)@@@")
```
make this a constant. never MustCompile in a function.

mfojtik commented:
@mnagy make the crash fix as a separate commit and follow the commit message format we use for importing commits from k8s

jhadvig commented on hack/util.sh:
```diff
@@ -74,9 +74,9 @@ function start_os_server {
     ps -ef | grep openshift
     echo "[INFO] Starting OpenShift server"
     sudo env "PATH=${PATH}" OPENSHIFT_PROFILE=web OPENSHIFT_ON_PANIC=crash openshift start \
-	 --master-config=${MASTER_CONFIG_DIR}/master-config.yaml \
```
undo this please

mnagy replied:
why?

mfojtik replied:
That change is not relevant to the topic of this PR. If somebody will be `git blame` this portion of code, your name will show up :-)

jhadvig commented on hack/util.sh:
```diff
@@ -556,6 +556,15 @@ function create_image_streams_extended {
     echo "[INFO] Registry IP - ${registry}"
 }
 
+# Make ourselves the default resolver. This is needes so FQDNs such as
```
s/needes/needed/

jhadvig commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,311 @@
+)
+
+// getPods looks up pods with "oc get" applying a label selector and returns
+// the names of matching pods
```
nuke 'the'

jhadvig commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,311 @@
+		case "MYSQL_MASTER_USER", "MYSQL_MASTER_PASSWORD":
+			// Ignore
+		default:
+			return nil, errors.New("Unknown environment variable named " + s[0])
```
use fmt.Errorf()
or check if there isn't something suitable in the k8s errors pkg

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,311 @@
+		Endpoint:  service + "." + oc.Namespace() + ".svc.cluster.local",
+	}
+
+	out, err := oc.Run("env").Args("pod", targetPod, "--list").Output()
```
don't rely on output formatting of `oc env`. if it change in future, this code will be broken.
If you want to get the current env vars, get the Pod using REST client directly.

jhadvig commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,311 @@
+	return out == "mysqld is alive", nil
+}
+
+// WaitUntilUp continuously waits for the server to respond to pings, up until timeout.
```
Does this method waits only for ping ? couldnt it depend on some other mechanism ?

jhadvig commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,311 @@
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo"
```
g "github.com/onsi/ginkgo"

jhadvig commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,311 @@
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
```
o "github.com/onsi/gomega"

mnagy commented:
PTAL

rhcarvalho commented on hack/util.sh:
```diff
@@ -517,7 +517,7 @@ function auth_setup_extended {
 	# Make oc use ${MASTER_CONFIG_DIR}/admin.kubeconfig, and ignore anything in the running user's $HOME dir
 	sudo chmod -R a+rwX "${ADMIN_KUBECONFIG}"
-	echo "[INFO] To debug: export ADMIN_KUBECONFIG=$ADMIN_KUBECONFIG"
+	echo "[INFO] To debug: export KUBECONFIG=$ADMIN_KUBECONFIG"
```
If this is to be fixed, you could fix it in a separate small PR?

mnagy commented:
The default bucket is gone after I rebased:
```
[ERROR] Extended test bucket default not found
```
[test][extended:images]

mnagy commented:
That didn't work, but I've triggered a manual rebuild in Jenkins with correct bucket set.

mnagy commented:
So the test was re-run, but the `images` bucket was just appended, so I cancelled the test, it'd break down.

liggitt commented:
You can restart now, I removed the extended default comment

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,266 @@
+	"k8s.io/kubernetes/pkg/util/wait"
+)
+
+func selector(selector string) labels.Selector {
```
why this is a func? it looks pretty generic to be special function... just call labels.Parse()

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,266 @@
+// getPods looks up pods with "oc get" applying a label selector and returns
+// names of matching pods
+func getPods(oc *exutil.CLI, label labels.Selector) (pods []string, err error) {
+	podList, err := oc.KubeREST().Pods(oc.Namespace()).List(label, nil)
```
do not pass `oc` just pass PodInterface

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,266 @@
+func pollForPods(oc *exutil.CLI, label labels.Selector, count int, timeout time.Duration) ([]string, error) {
+	var pods []string
+	err := wait.Poll(1*time.Second, timeout, func() (bool, error) {
+		p, e := getPods(oc, label)
```
same here, don't pass CLI, just PodInterface (see the builds test)

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,266 @@
+	Env     map[string]string
+}
+
+func NewMysql(oc *exutil.CLI, podName, service string) (*mysql, error) {
```
godoc

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,266 @@
+	if err != nil {
+		return nil, err
+	}
+	mysql.Env = make(map[string]string, len(pod.Spec.Containers[0].Env))
```
I don't understand what are you doing here... can you describe this in godoc?

mnagy replied:
It just makes the environment variables available as a simple map. Yes, I'll mention it in godoc.

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,266 @@
+
+// WaitForMySQL pings the MySQL server
+func (m *mysql) WaitForMySQL() (bool, error) {
+	out, err := m.oc.ExecWithoutExit(m.PodName, "mysqladmin -h "+"127.0.0.1"+" -u"+"root"+" ping").Output()
```
use fmt.Sprintf()

rhcarvalho replied:
Wait wait.. this should be a single string... no need for `fmt` nor concatenation:
```go
"mysqladmin -h 127.0.0.1 -uroot ping"
```

mfojtik replied:
yeah

mfojtik replied:
Also this should be a standalone function, don't bundle this into *mysql struct

mnagy replied:
Right, sorry about that. It's a leftover from before, will fix.

mnagy replied:
But I'll rewrite rest of string concatenations to use fmt.Sprintf.

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,266 @@
+}
+
+// TODO: Generalize, move to exutil
+func (m *mysql) QueryAsRoot(sql string) (string, error) {
```
godoc

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,266 @@
+}
+
+// TODO: Generalize, move to exutil
+func (m *mysql) QueryAsUser(sql string) (string, error) {
```
godoc

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,266 @@
+}
+
+// TODO: Generalize, move to exutil
+func (m *mysql) WaitForMySQLOutput(timeout time.Duration, root bool, sql, resultSubstr string) error {
```
godoc

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,266 @@
+	)
+
+	g.Describe("MySQL replication", func() {
+		g.JustBeforeEach(func() {
```
this should be a function... I think already made a comment about that here...

InitializeNFSVolumes() ?

mnagy replied:
Yes, that's why I created the CreatePersistentVolume function. BTW, we don't use nfs anymore. But yeah, I guess I'll move that as well.

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,266 @@
+			o.Expect(err).NotTo(o.HaveOccurred())
+		})
+
+		g.It("should work", func() {
```
change this

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,266 @@
+		})
+
+		g.AfterEach(func() {
+			err := oc.AsAdmin().Run("delete").Args("pods", "--all").Verbose().Execute()
```
you can just delete everything in the namespace (oc delete all --all -n NAMESPACE)

rhcarvalho replied:
You can do this in one call to `oc` (and using the namespace is a good idea, can avoid hard to debug problems later):

```go
err = oc.AsAdmin().Run("delete").Args("all,pvc,pv", "--all", "-n", namespace).Verbose().Execute()
```

rhcarvalho replied:
To be clear, you can do a single `oc delete all,pvc,pc --all -n <namespace` instead of these 4 calls here.

mnagy replied:
This was a relic since the nfs.. Deleting PVs before the pods was very troublesome. Will shorten it.

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,266 @@
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			// Get pod names
+			podNames, err := getPods(oc, selector("name=mysql-master"))
```
if this returns just names for pods, the function should be called "getPodsNames"

mfojtik commented on test/extended/util/framework.go:
```diff
@@ -145,6 +146,35 @@ func CreatePodForImage(dockerImageReference string) *kapi.Pod {
+			},
+			AccessModes: []kapi.PersistentVolumeAccessMode{
+				kapi.ReadWriteOnce,
+				// TODO: Uncomment after https://github.com/kubernetes/kubernetes/pull/10833 is merged
```
it is merged

mnagy replied:
not in OS :(

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,266 @@
+			err = oc.Run("env").Args("dc", "mysql-master", "MYSQL_ROOT_PASSWORD=newpass").Execute()
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			podNames, err = pollForPods(oc, selector("deployment=mysql-master-2"), 1, 60*time.Second)
```
this should be "waitForPods" with ConditionFunc.

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,266 @@
+}
+
+type mysql struct {
+	oc      *exutil.CLI
```
I'm OK with creating helper structs like this, but don't like bundling the exutil.CLI here. Is there a way to make this more generic? And don't bundle *exutil.CLI ? You really just need PodInterface here and not the full exutil.CLI

mnagy commented:
Comments addressed, PTAL.

jhadvig commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,148 @@
+			g.By("creating a table in master")
+			out, err := master.QueryAsUser("CREATE TABLE tbl (col1 VARCHAR(20), col2 VARCHAR(20));")
+			o.Expect(err).NotTo(o.HaveOccurred())
+			fmt.Printf("Output: \"%s\"\n", out)
```
Is printing the output necessary ? if not, :boom: it

jhadvig commented on test/extended/util/framework.go:
```diff
@@ -160,3 +268,14 @@ func ExtendedTestPath() string {
 	return filepath.Join(append([]string{ExtendedTestPath()}, elem...)...)
 }
+
+// MustParseLabelsOrDie turns the given string itno a label selector or
```
s/itno/into

jhadvig commented on test/extended/util/framework.go:
```diff
@@ -160,3 +268,14 @@ func ExtendedTestPath() string {
+// MustParseLabelsOrDie turns the given string itno a label selector or
+// panics; for tests or other cases where you know the string is valid.
+// TODO: Move this to the upstream labels package.
+func MustParseLabelsOrDie(str string) labels.Selector {
```
:+1: for the name :)

mfojtik replied:
ParseLabelsOrDie ... there is no need for Must (and when we talked trough this I told you... shorter is better).

jhadvig commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,148 @@
+	kapi "k8s.io/kubernetes/pkg/api"
+)
+
+var _ = g.Describe("images: MySQL replication template", func() {
```
I guess that the main `Describe` should be more descriptive.

mnagy replied:
It's analogous to what we have in mysql_ephemeral.go and I'm not sure if I can think of anything better, because the one thing that is covered by this tests really is the replication template (I was thinking of MySQL image, but that's not true, since we have mysql_ephemeral tests). I'm open to suggestions though.

jhadvig commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,148 @@
+			err = oc.KubeFramework().WaitForAnEndpoint("mysql-slave")
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			// Get pod names
```
The method below is descriptive enough. Nuke this comment

mnagy commented:
Rebased, addressed comments, added a lot more cases (restarting both master and slave) and tidied up. Still very few nits remain, plan on doing that tomorrow.

mnagy commented:
[test][extended:core]

Would be nice if we had a way to pass along a focus argument.

mfojtik commented on test/extended/images/db_helpers.go:
```diff
@@ -0,0 +1,120 @@
+package images
```
move this file into util package, call it "mysql_image_helpers.go"

mnagy replied:
Why? Those functions will not be used by any other tests. This is more in line with what we have for networking (networking/util.go). Also, I wanted the name to be more generic so we can add PostgreSQL later and maybe share functionality (though maybe sql_helpers.go would be a better name).

mfojtik replied:
this helpers will be available for every test file in "images" directory. If we want to call them from a different place (for example testing some example application), you have to import the image tests which is not what you want, right?

mnagy replied:
I'm not sure why you would want that, but I suppose. I'll move it then.

mfojtik commented on test/extended/images/db_helpers.go:
```diff
@@ -0,0 +1,120 @@
+
+// MySQL is a MySQL helper for executing commands
+type MySQL struct {
+	oc        *exutil.CLI
```
nuke this

mfojtik commented on test/extended/images/db_helpers.go:
```diff
@@ -0,0 +1,120 @@
+// variables like username and password for easier use.
+func NewMysql(oc *exutil.CLI, podName, masterPodName string) (*MySQL, error) {
+	m := &MySQL{
+		oc:      oc,
```
nuke this

mfojtik replied:
replace this with PodInterface

mfojtik commented on test/extended/images/db_helpers.go:
```diff
@@ -0,0 +1,120 @@
+		PodName: podName,
+	}
+
+	if pod, err := oc.KubeREST().Pods(oc.Namespace()).Get(podName); err != nil {
```
you just need PodInterface to get the container name. No need to have full CLI here.

mfojtik commented on test/extended/images/db_helpers.go:
```diff
@@ -0,0 +1,120 @@
+
+// GetEnv will get and save environment variables from a pod
+func (m *MySQL) GetEnv(podName string) error {
+	pod, err := m.oc.KubeREST().Pods(m.oc.Namespace()).Get(podName)
```
same here, oc.KubeREST().Pods(namespace) gives you PodInterface, use this instead of CLI.

mfojtik commented on test/extended/images/db_helpers.go:
```diff
@@ -0,0 +1,120 @@
+}
+
+// WaitForMySQL pings the MySQL server
+func (m *MySQL) WaitForMySQL() (bool, error) {
```
move this away from MySQL struct, make it a standalone helper. podname, containername and oc as arguments.

jhadvig replied:
@mfojtik +1
We need to make this function more abstract so it can be used by other DBs as well

mfojtik commented on test/extended/images/db_helpers.go:
```diff
@@ -0,0 +1,120 @@
+}
+
+// QueryAsUser executes an SQL query as a root user and returns the result.
+func (m *MySQL) QueryAsRoot(sql string) (string, error) {
```
Make this two standalone funcs:

MySQLQueryAsRoot(*CLI, *MySQL, query)
MySQLQueryAsUser(...)

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,143 @@
+		g.JustBeforeEach(func() {
+			var err error
+			persistentVolumes, err = exutil.SetupHostPathVolumes(oc.AdminKubeREST().PersistentVolumes(), "mysql-replication", "512Mi", 1)
+			o.Expect(err).NotTo(o.HaveOccurred())
```
I think you also want to check if persistentVolumes is not empty, right?

mnagy replied:
no need for that

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,143 @@
+			o.Expect(err).NotTo(o.HaveOccurred())
+		})
+
+		g.AfterEach(func() {
```
this will run more often as you think, make sure it is what you want ;-)

mnagy replied:
It's ok for now, though I'm going to change it because I'll be adding another template.

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,143 @@
+				Args(
+				"-f", templatePath,
+				"-v", "MYSQL_USER="+user+",MYSQL_PASSWORD="+userPass+",MYSQL_ROOT_PASSWORD="+rootPass+",MYSQL_DATABASE="+database).
+				OutputToFile("config.json")
```
You can create the template in OpenShift and then instantiate it without need of storing temporary file (I hope). Can you check if that approach is doable?

mnagy replied:
Apparently, new-app can take the json template file directly.

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,143 @@
+			err = oc.Run("create").Args("-f", configFile).Execute()
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			createHelpers := func(masterDeployment, slaveDeployment string, slaveCount int) (*MySQL, []*MySQL) {
```
extract this to util/helpers package with proper godoc?

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,143 @@
+				return master, slaves
+			}
+
+			assertReplicationIsWorking := func(master *MySQL, slaves []*MySQL) {
```
same here, don't hide this function if you don't have strong reason to do that... just create regular helper function in mysql_helpers.go

mnagy replied:
I'll move the one that creates the helpers, but keep this one more or less as it is. It is what ginkgo documentation recommends: https://onsi.github.io/ginkgo/#shared-example-patterns

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,143 @@
+			}
+
+			assertReplicationIsWorking := func(master *MySQL, slaves []*MySQL) {
+				table := fmt.Sprintf("table_%s", rand.String(10))
```
You can use the namespace here oc.Namespace()

mnagy replied:
Why? You need a new random table name for each invocation of this function.

mnagy replied:
Though I was thinking of saving each new table into the array and then at the end check if all tables are present, to make sure persistence is working correctly.

mfojtik commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,143 @@
+			g.By("after slave is scaled to 0 and then back to 4 replicas")
+			err = oc.Run("scale").Args("dc", "mysql-slave", "--replicas=0").Execute()
+			o.Expect(err).NotTo(o.HaveOccurred())
+			err = oc.Run("scale").Args("dc", "mysql-slave", "--replicas=4").Execute()
```
you want to make sure you have 4 replicas by now, this smells flaky.

mnagy replied:
The createHelpers() is making sure there are four of them.

mfojtik commented on test/extended/util/cli.go:
```diff
@@ -66,6 +82,13 @@ func (c *CLI) Username() string {
 }
 
+// AsAdmin changes current config file path to the admin config.
+func (c *CLI) AsAdmin() *CLI {
```
AsAnAdmin? ;-)

mnagy replied:
I think the 'An' shouldn't be there. @nhr?

jhadvig replied:
As an admin, should be correct

mnagy replied:
Shouldn't we therefore have AsAUser? I don't like it. @bparees thoughts?

mfojtik replied:
AsAUser is default. Having to run things as admin is "special" case.

rhcarvalho replied:
Bikeshedding alarm on? Come on guys, let's focus on what is needed to merge this... I don't want to see @mnagy working on this until Christmas :santa:

bparees replied:
+1 @rhcarvalho.  AsAdmin is fine.  AsAnAdmin is also fine, though it looks worse.

mfojtik commented on test/extended/util/cli.go:
```diff
@@ -226,6 +258,17 @@ func (c *CLI) Args(args ...string) *CLI {
 
+// ExecWithoutExit executes a command in container ignoring non-zero exit value.
+// TODO: Remove this when https://github.com/kubernetes/kubernetes/pull/13728 is merged in OS
+func (c *CLI) ExecWithoutExit(pod, container, cmd string) *CLI {
```
I don't think this is needed anymore... See BehaviorOnFatal() in k8s.

mnagy replied:
I'm aware of it, but it was not merged into OS yet.

mfojtik replied:
it was :-)

mnagy replied:
Are you sure? I have pulled my master and doing `git grep BehaviorOnFatal` produces nothing.

mfojtik replied:
oh sorry, it is pending https://github.com/openshift/origin/pull/4611 I have to rework that.

mfojtik replied:
same as above, @smarterclayton PR will make oc.Run("exec") not exit

mfojtik commented on test/extended/util/framework.go:
```diff
@@ -223,6 +226,38 @@ var CheckDeploymentFailedFunc = func(d *kapi.ReplicationController) bool {
 }
 
+// GetRunningPodNames looks up pods that are in a Running state and returns their names.
+func GetRunningPodNames(c kclient.PodInterface, label labels.Selector) (podNames []string, err error) {
```
Make this generic function that accept a ConditionFunc as argument. Then use the ConditionFunc to determine what pods to return....

GetPodNamesByFilter(PodInterface, labels, conditionFunc).

mfojtik commented on test/extended/util/framework.go:
```diff
@@ -265,6 +300,79 @@ func CreatePodForImage(dockerImageReference string) *kapi.Pod {
+
+// SetupHostPathVolumes will create multiple PersistentVolumes with given capacity
+func SetupHostPathVolumes(c kclient.PersistentVolumeInterface, prefix, capacity string, count int) (volumes []*kapi.PersistentVolume, err error) {
+	if count < 1 {
```
this check should be done when you call this function?

mnagy replied:
what do you mean?

mfojtik replied:
remove the if and move it to the place where you call this function. Make sure you call it with count < 1 ? I believe this is just checking you are not calling this with count == 0, right? Then just godoc that fact here and check the integer in place where you call this.

mfojtik commented on test/extended/util/framework.go:
```diff
@@ -265,6 +300,79 @@ func CreatePodForImage(dockerImageReference string) *kapi.Pod {
+		return volumes, err
+	}
+	for i := 0; i < count; i++ {
+		dir, err := ioutil.TempDir(rootDir, fmt.Sprintf("%0.4d", i))
```
I have bad feeling that you will also need to chcon these directories in order to make them work with OpenShift and SELinux

mnagy replied:
@csrwng: @mfojtik mentioned that you might now the correct labels to apply so you can host mount the directory?

bparees replied:
@mnagy probably chcon -Rt svirt_sandbox_file_t <file>

csrwng replied:
yup, that's what we do in our own e2e https://github.com/openshift/origin/blob/master/hack/test-end-to-end-docker.sh#L28

jhadvig commented on test/extended/util/cli.go:
```diff
@@ -35,10 +39,22 @@ type CLI struct {
+	ExitStatus int
+}
+
+func (e ExitError) Error() string {
```
Add godoc if we decide to stick to this method

mnagy commented:
PTAL

mfojtik commented on test/extended/util/cli.go:
```diff
@@ -36,10 +40,23 @@ type CLI struct {
+// ExitError happens when a command running inside of a container launched
+// using oc exec exits with non-zero code.
+// TODO: Remove this when https://github.com/kubernetes/kubernetes/pull/13728 is merged in OS
+type ExitError struct {
```
https://github.com/openshift/origin/pull/4771

when this gets merged, you will not need this.

mfojtik commented on test/extended/util/cli.go:
```diff
@@ -166,6 +190,15 @@ func (c *CLI) KubeREST() *kclient.Client {
 }
 
+// AdminKubeREST provides a Kubernetes REST client for the admin user.
+func (c *CLI) AdminKubeREST() *kclient.Client {
```
@smarterclayton just added this in his PR ;-)

mfojtik commented:
LGTM, but you really want to incorporate @smarterclayton changes before merging this.

mfojtik commented:
@mnagy Clayton PR is now merging, so feel free to pick it up for this PR.

mnagy commented:
Rebased and added the SELinux chcon call.

mnagy commented:
Rebased and fixed race with pod deletion.

bparees commented on test/extended/util/framework.go:
```diff
@@ -267,6 +322,80 @@ func CreatePodForImage(dockerImageReference string) *kapi.Pod {
+	prefix = fmt.Sprintf("%s%s-", pvPrefix, prefix)
+	for _, pv := range pvs.Items {
+		if strings.HasPrefix(pv.Name, prefix) {
+			c.Delete(pv.Name)
```
does this actually delete the path on the host?

mnagy replied:
No. Do you think it should? I wanted to avoid that for debugging purposes, plus the files are created in temporary directory.

bparees replied:
yes, tests should clean up after themselves.  we don't want to fill up people's machines if they run the test repeatedly.

mnagy replied:
Ok, will fix.

rhcarvalho replied:
Not sure if this would be helpful here as I'm not familiar with the context, but it came to mind that we can have OpenShift cleanup PVs for us:
https://docs.openshift.org/latest/admin_guide/persistent_storage_nfs.html#provisioning

```json
 "persistentVolumeReclaimPolicy": "Recycle"
```

bparees replied:
that would empty the dir but not delete the dir, right?  which is probably good enough here, admittedly.

but only if the openshift server sticks around long enough to do the recycling, which is probably not guaranteed especially in the face of a test failure.

mnagy replied:
So, the files get deleted once extended tests are re-run. So you can still debug the tests after they are run, but you won't get clutter growing after multiple test runs.

bparees commented:
lgtm.  as long as it passes [test][extended:core]

mnagy commented:
@bparees: Tests are apparently failing because of something else and the extended tests that I'm touching in this PR are passing as well. Look for "MySQL replication template" in Jenkins logs.

bparees commented:
[merge]

php-coder commented on test/extended/images/mysql_replica.go:
```diff
@@ -0,0 +1,147 @@
+)
+
+// CreateMySQLReplicationHelpers creates a set of MySQL helpers for master,
+// slave an en extra helper that is used for remote login test.
```
Looks like a typo: `an en` should we replace it by `and an`?
